Add binary data for the XCTS equations

nilsvu opened the pull request:
## Proposed changes

This class provides the background data for the XCTS equations to solve for compact-object binaries. It essentially superposes two isolated solutions, e.g. two Kerr solutions in Kerr-Schild coordinates to obtain SKS initial data.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Add missing dt(K) implementation for Schwarzschild

nilsvu committed:
Add a missing tmpl::pin

nilsvu committed:
Add inverse conformal metric to common vars

nilsvu committed:
Add binary data for the XCTS equations

moxcodes commented on src/PointwiseFunctions/AnalyticData/Xcts/Binary.cpp:
```diff
@@ -0,0 +1,111 @@
+}
+
+template <typename DataType>
+void BinaryVariables<DataType>::operator()(
```
This has the appearance of being a generic calculation that doesn't much depend on the details of the binary solution. Does it make sense to move it to the `CommonVariables`?

nilsvu replied:
I considered that, but it requires all solutions implement `deriv(ShiftBackground)` _and_ `dt(ConformalMetric)` instead. In most cases it's easier to just set the longitudinal-shift-minus-dt(gamma) to zero. In fact, I can't think of another scenario right now where it's non-zero. That said, it may turn out to be useful to move this to `CommonVariables` at some point in the future, but not right now I think.

moxcodes commented on src/PointwiseFunctions/AnalyticData/Xcts/Binary.hpp:
```diff
@@ -0,0 +1,344 @@
+  Binary& operator=(Binary&&) = default;
+  ~Binary() = default;
+
+  Binary(std::array<double, 2> xcoords,
```
take non-fundamental args by const ref or move them into their member variables.

nilsvu replied:
I got a clang-tidy complaint here because I think `std::array<double, 2>` is trivially-copyable :)

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/Xcts/Binary.hpp:
```diff
@@ -0,0 +1,344 @@
+template <typename IsolatedObjectRegistrars,
+          typename Registrars = tmpl::list<
+              Xcts::AnalyticData::Registrars::Binary<IsolatedObjectRegistrars>>>
+class Binary : public ::AnalyticData<3, Registrars> {
```
I can't find the documentation for this class. I'm not sure if that's an issue with how I rendered the PR docs. Could you double check that it shows up for you? Maybe you need to `cond` the forward declaration on line 135?

nilsvu replied:
It shows up on my machine with Doxygen 1.9.1, but didn't in the rendered docs from the CI workflow. I added the `cond`, maybe it works now :)

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/Xcts/Binary.cpp:
```diff
@@ -0,0 +1,111 @@
+  for (size_t i = 0; i < 3; ++i) {
+    for (size_t j = 0; j < 3; ++j) {
+      for (size_t k = 0; k <= j; ++k) {
+        for (size_t alpha = 0; alpha < 2; ++alpha) {
```
maybe rename `alpha` to `object_index` since alpha is used for the lapse, and Greek indices are sometimes used in GR?

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticData/Xcts/Binary.py:
```diff
@@ -0,0 +1,40 @@
+import numpy as np
+
+centers = [-5., 6.]
+masses = [1., 0.43]
```
maybe `1.1` or something not exactly one, just to catch a possible typo?

nilsdeppe reviewed (changes requested):
You can squash immediately

nilsdeppe reviewed (dismissed):
The documentation is showing up now :) Not sure if it was the cond or if I had done something wrong last time :shrug:

nilsvu committed:
Support remaining tags in XCTS Flatness solution

nilsvu committed:
Support conformal matter scale in Schwarzschild solution

nilsvu commented:
Rebased on develop to fix the formatting CI failure. Also added the `Tags::Conformal` prefix to the matter tags to work with #3107. I took the two commits "Support conformal matter scale in Schwarzschild solution" and "Support remaining tags in XCTS Flatness solution" from #3107 for this purpose. Doesn't really matter which PR merges those I think.

nilsvu commented:
@moxcodes would you be willing to review this one?

nilsdeppe commented:
This fails all builds

nilsvu commented:
Oh no.. squashed in a missing include.

moxcodes commented on src/PointwiseFunctions/AnalyticData/Xcts/Binary.hpp:
```diff
@@ -0,0 +1,350 @@
+  }
+
+  std::array<double, 2> x_coords() const noexcept { return xcoords_; }
+  std::array<std::unique_ptr<IsolatedObjectBase>, 2> superposed_objects()
```
I'm confused about what happens with the return type here -- does it try to copy the `unique_ptr`s?
I can't find anywhere this function is used -- are these accessor functions necessary? If they're needed, then it seems like this function needs to do some kind of `get_clone` trick or to return the contents by const reference.

nilsvu replied:
Good catch, I switched these accessors to return const-refs and added a small test.

moxcodes commented on tests/Unit/PointwiseFunctions/AnalyticData/Xcts/Test_Binary.cpp:
```diff
@@ -0,0 +1,111 @@
+    Tags::ConformalFactor<DataVector>>;
+
+template <typename IsolatedObjectRegistrars>
+struct BinaryProxy {
```
I think it's probably best to avoid using the term `proxy` outside of referring to the charm construct.

nilsvu replied:
Agreed, though this "proxy" business is done in all other analytic solution tests (for all systems) so I think it's better to be consistent in this PR. But would be good to rename them.

moxcodes commented:
(feel free to squash any changes immediately from my review)

nilsvu commented:
Rebased and squashed in the changes above. Thanks for reviewing @nilsdeppe and @moxcodes !

nilsdeppe commented:
@nilsleiffischer looks like something went wrong. Most of the builds are failing

nilsvu commented:
Ah I missed this interaction with #3068 in the rebase. I squashed this diff into the commit "Add inverse conformal metric to common vars":

```diff
diff --git a/src/PointwiseFunctions/AnalyticSolutions/Xcts/Kerr.hpp b/src/PointwiseFunctions/AnalyticSolutions/Xcts/Kerr.hpp
index 16f823217f..01226fde0e 100644
--- a/src/PointwiseFunctions/AnalyticSolutions/Xcts/Kerr.hpp
+++ b/src/PointwiseFunctions/AnalyticSolutions/Xcts/Kerr.hpp
@@ -34,7 +34,6 @@ using KerrVariablesCache = cached_temp_buffer_from_typelist<
     tmpl::push_back<
         common_tags<DataType>,
         Tags::ConformalMetric<DataType, 3, Frame::Inertial>,
-        Tags::InverseConformalMetric<DataType, 3, Frame::Inertial>,
         ::Tags::deriv<Tags::ConformalMetric<DataType, 3, Frame::Inertial>,
                       tmpl::size_t<3>, Frame::Inertial>,
         gr::Tags::TraceExtrinsicCurvature<DataType>,
```
